File: src/commands/push.ts
import { parseFlag, parseTrack } from "../args";
import { resolveRepoRoot } from "../repo-root";
import { resolveRootDir } from "../root";
import { getDefaultTrack } from "../sync-config";
import * as siblingPush from "../tracks/sibling/push";
import * as shadowPush from "../tracks/shadow/push";

export function run(args: string[]): void {
  const { value: rootFlag } = parseFlag(args, "root");
  const repoRoot = resolveRepoRoot();
  const rootDir = resolveRootDir(repoRoot, rootFlag);

  let track, rest;
  try {
    ({ track, rest } = parseTrack(args, getDefaultTrack(repoRoot, rootDir)));
  } catch (err) {
    throw new Error(`push: ${(err as Error).message}`);
  }
  if (track === "sibling") return siblingPush.run(rest);
  if (track === "shadow") return shadowPush.run(rest);
}

File: go/internal/commands/push.go
package commands

import (
	"fmt"
	"os"

	"plan-sync/go/internal/args"
	"plan-sync/go/internal/tracks/sibling"
)

const pushHelp = `Usage: plan-sync push [--track <sibling|shadow>] [--root <dir>]

Pushes manifest-listed files to the sync destination (sibling repo or
shadow git ref).

Flags:
  --track <sibling|shadow>  Sync track (optional if a default track is persisted)
  --root <dir>              Root directory to sync (optional, e.g. ".omc")
`

// Push is the Go port of src/commands/push.ts.
func Push(argv []string) error {
	if args.HasHelpFlag(argv) {
		fmt.Fprint(os.Stdout, pushHelp)
		return nil
	}

	repoRoot, rootDir, err := resolveContext(argv)
	if err != nil {
		return err
	}

	track, rest, err := resolveTrack("push", argv, repoRoot, rootDir)
	if err != nil {
		return err
	}

	switch track {
	case args.TrackSibling:
		return sibling.Push(rest)
	case args.TrackShadow:
		return ShadowPush(rest)
	}
	return nil
}
